clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -588,12 +588,6 @@ bool EmitAssemblyHelper::AddEmitPasses(legacy::PassManager &CodeGenPasses,
   // this also adds codegenerator level optimization passes.
   CodeGenFileType CGFT = getCodeGenFileType(Action);
 
-  // Add ObjC ARC final-cleanup optimizations. This is done as part of the
```

Comment:
@cachemeifyoucan As far as I can tell, it's already the case that for non-LTO builds, this pass is only added for optimization level > 0. Is that ok to do for all builds?

I checked and directly providing clang with IR that contains `llvm.objc.clang.arc.use` and compiling it with `-O0` fails instruction selection as a result of not running this pass, so I'm assuming these intrinsics are only introduced at `-O1` or higher.

Reply:
@ahatanak probably know better. I think when the arc intrinsics are present, the pass is mandatory but I am not sure if the intrinsics are only produced for -O1 and higher.

Reply:
Don't run this optimization on O0 has been there for 10+ yrs since 1170b0885f635ef8566. Unfortunately I don't know the committer, but he/she/they probably did it for a reason :(

Reply:
The intrinsic calls are needed to prevent ARC optimizer from moving releases to an earlier point that can cause objects to be deallocated prematurely. clang doesn't emit the intrinsic calls if `OptimizationLevel` is zero.